### PR TITLE
Generate: store special token tensors under a unique variable name

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

See this comment in the end-to-end generation PR: https://github.com/huggingface/transformers/pull/30788#issuecomment-2228710326

## Problem TL;DR
Compiled code can't create tensors as `x=torch.tensor(x)` (tensor output var name = non-tensor input var name), the graph overwrites the argument (non-tensor) with the output (tensor), causing the graph to be missing the input to this node the 2nd time it is called (because it was overwritten). It is a [known limitation](https://pytorch.org/docs/stable/torch.compiler_cudagraph_trees.html#limitations) of `torch.compile`.

Our code that converts the special tokens into tensors falls into this pattern.

## Solution

Write the special tokens converted into tensors under a new variable name. A bit annoying, we will now have e.g. `pad_token_id` (integer) and `pad_token_tensor` (tensor) in the `generation_config` object throughout `generate`.

Note: We can't change the variable name where these tokens are read from, which would be much cleaner. This is because: 
1. in end-to-end compilation we can't deepcopy  
2. the attribute is set outside `generate` (at `GenerationConfig` creation time)

as such, even if we create an auxiliary attribute with a different name to read from, the original attribute will be overwritten, leading to the original issue 😢

### CODE REVIEW

File: src/transformers/generation/utils.py [hunk 1]
```diff
@@ -1389,6 +1392,16 @@ def _prepare_generation_config(
...
+            # If `generation_config` is provided, let's fallback ALL special tokens to the default values for the model
+            if not using_model_generation_config:
+                if generation_config.bos_token_id is None:
+                    generation_config.bos_token_id = self.generation_config.bos_token_id
+                if generation_config.eos_token_id is None:
+                    generation_config.eos_token_id = self.generation_config.eos_token_id
+                if generation_config.pad_token_id is None:
+                    generation_config.pad_token_id = self.generation_config.pad_token_id
+                if generation_config.decoder_start_token_id is None:
+                    generation_config.decoder_start_token_id = self.generation_config.decoder_start_token_id
```
Comment (gante) [inline src/transformers/generation/utils.py:1404]: This is equivalent to the changes in [this PR](https://github.com/huggingface/transformers/pull/31254), which are better suited to this function -- handling retrocompatibility wrt config files

File: tests/generation/test_utils.py [hunk 2]
```diff
@@ -3196,6 +3196,39 @@ def test_assisted_decoding_in_gpu_cpu(self):
         )
         self.assertTrue(input_length <= out.shape[-1] <= input_length + 20)
 
+    def test_special_tokens_fall_back_to_model_default(self):
```
Comment (gante) [inline tests/generation/test_utils.py:3199]: This test was missing in https://github.com/huggingface/transformers/pull/31254 😛

File: src/transformers/generation/utils.py [hunk 3]
```diff
@@ -1540,21 +1547,26 @@ def _tensor_or_none(token_kwargs, token_self, device=None):
...
-        generation_config.eos_token_id = eos_token_id
-        generation_config.pad_token_id = pad_token_id
-        generation_config.decoder_start_token_id = decoder_start_token_id
+        # NOTE: this must be written into a different attribute name than the one holding the original special tokens
+        # (in their non-tensor form), in order to enable end-to-end compilation. See
+        # https://pytorch.org/docs/stable/torch.compiler_cudagraph_trees.html#limitations
+        generation_config.bos_token_tensor = bos_token_tensor
+        generation_config.eos_token_tensor = eos_token_tensor
+        generation_config.pad_token_tensor = eos_token_tensor
+        generation_config.decoder_start_token_tensor = decoder_start_token_tensor
```
Comment (ArthurZucker) [inline src/transformers/generation/utils.py:1569]: would it work to have a property, eos_token_id with _eos_token_tensor underlying? When you set it you cast to tensor format. might be simpler in general ?

**Review by ArthurZucker — APPROVED** [review]
Wow nice finding!

File: src/transformers/generation/utils.py (hunk 3 above)
Comment (gante) [reply src/transformers/generation/utils.py:1569]: uhmmm `@property` will further tangle us with state/python classes, which I'm not a fan of for compile purposes 🤔 

I am going to rename the tensor variables from `xxx_token_tensor` to `_xxx_token_tensor` though, to help with readability!

File: src/transformers/models/musicgen/modeling_musicgen.py [hunk 4]
```diff
@@ -1539,75 +1539,43 @@ def generate(
         logits_processor = logits_processor if logits_processor is not None else LogitsProcessorList()
         stopping_criteria = stopping_criteria if stopping_criteria is not None else StoppingCriteriaList()
 
-        if generation_config.pad_token_id is None and generation_config.eos_token_id is not None:
```
Comment (gante) [inline src/transformers/models/musicgen/modeling_musicgen.py:1542]: some more musicgen standardization :)

Comment (gante) [reply src/transformers/models/musicgen/modeling_musicgen.py:1542]: confirmed: slow tests have the same failures as on `main`

**Review by zucchini-nlp — APPROVED** [review]
An interesting behavior from `compile`. Thanks for handling!

**Pull request merged**